Woo REST API: Migrate OrderRestClient

hafizrahman opened the pull request:
Closes https://github.com/woocommerce/woocommerce-android/issues/8013

This PR migrates `OrderRestClient` to use `WooNetwork`.

### Testing
As this is a migration, please mostly check the code and ensure no functions are missed, and that the parameters and used HTTP methods are correct. Also ensure related tests still pass.

hafizrahman committed:
Add network dependencies.

hafizrahman committed:
Small var -> val fix.

hafizrahman committed:
Fix mocked test for WCOrdersTest.

hafizrahman committed:
Refactor fetchOrder()

hafizrahman committed:
Refactor fetchOrderListSummaries()

hafizrahman committed:
Refactor fetchOrdersByIds()

hafizrahman committed:
Refactor fetchOrderStatusOptions()

hafizrahman committed:
Refactor searchOrders()

hafizrahman committed:
Refactor fetchSingleOrder()

hafizrahman committed:
Refactor fetchOrderCount()

hafizrahman committed:
Refactor fetchHasOrders()

hafizrahman committed:
Refactor updateOrder()

hafizrahman committed:
Refactor fetchOrderNotes()

hafizrahman committed:
Refactor postOrderNote()

hafizrahman committed:
Refactor postOrderNote()

hafizrahman committed:
Refactor addOrderShipmentTrackingForOrder()

hafizrahman committed:
Refactor deleteShipmentTrackingForOrder()

hafizrahman committed:
Refactor fetchOrderShipmentProviders()

hafizrahman committed:
Refactor createOrder()

hafizrahman committed:
Refactor updateOrder()

hafizrahman committed:
Refactor deleteOrder()

hafizrahman committed:
Various automated code reformatting.

hafizrahman committed:
Remove unneeded property.

hafizrahman committed:
Remove mentions of Jetpack tunnel in the comments as they're no longer strictly accurate.

hafizrahman commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCOrdersTest.kt:
```diff
@@ -140,7 +141,7 @@ class MockedStack_WCOrdersTest : MockedStack_Base() {
     fun testSearchOrdersError() {
         interceptor.respondWithError("jetpack-tunnel-root-response-failure.json")
-        orderRestClient.searchOrders(SiteModel(), "", 0)
+        orderRestClient.searchOrders(siteModel, "", 0)
```
Using `SiteModel()` crashed this instead because the origin value is zero.

hafizrahman commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCOrdersTest.kt:
```diff
@@ -170,7 +171,7 @@ class MockedStack_WCOrdersTest : MockedStack_Base() {
     fun testOrderListFetchError() {
         interceptor.respondWithError("jetpack-tunnel-root-response-failure.json")
-        orderRestClient.fetchOrders(SiteModel(), 0)
+        orderRestClient.fetchOrders(siteModel, 0)
```
Same as above.

hafizrahman committed:
Shortening Success/Error texts.

hafizrahman committed:
Spacing fixes.

hafizrahman committed:
OrderRestClient no longer needs to be of BaseWPComRestClient`

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -54,30 +46,27 @@ import org.wordpress.android.fluxc.store.WCOrderStore.OrderErrorType.GENERIC_ERR
 
-@Suppress("LargeClass")
+@Suppress("LargeClass", "TooManyFunctions")
 @Singleton
```
We can probably remove the `@Singleton` annotation here to align with the rest of the migrations.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -560,21 +560,22 @@ class OrderRestClient @Inject constructor(
+
             }
-            is JetpackError -> WooPayload(response.error.toWooError())
+            is WPAPIResponse.Error -> WooPayload(response.error.toWooError())
```
The `toWooPayload` would help simplify this code.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -597,27 +597,27 @@ class OrderRestClient @Inject constructor(
                 )
             }
-            is JetpackError -> WooPayload(response.error.toWooError())
+            is WPAPIResponse.Error -> WooPayload(response.error.toWooError())
```
Same as above, the `toWooPayload` handles also the `null` response.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -684,34 +684,33 @@ class OrderRestClient @Inject constructor(
+            site = site,
+            path = url,
+            clazz = OrderShipmentTrackingApiResponse::class.java,
+            body = params
```
np, it would be clearer to rename the variable to `body`.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -828,25 +828,26 @@ class OrderRestClient @Inject constructor(
-            )
+                )
+            }
+            is WPAPIResponse.Error -> WooPayload(response.error.toWooError())
```
Same as above comments, using `toWooPayload` would simplify the code.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -859,25 +859,26 @@ class OrderRestClient @Inject constructor(
-            )
+                )
+            }
+            is WPAPIResponse.Error -> WooPayload(response.error.toWooError())
```
Same

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -889,17 +887,16 @@ class OrderRestClient @Inject constructor(
         return when (response) {
-            is JetpackError -> WooPayload(response.error.toWooError())
-            is JetpackSuccess -> WooPayload(Unit)
+            is WPAPIResponse.Success -> WooPayload(Unit)
+            is WPAPIResponse.Error -> WooPayload(response.error.toWooError())
```
Same.

hichamboushaba commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:
```diff
@@ -21,7 +21,8 @@ import org.wordpress.android.fluxc.model.order.UpdateOrderRequest
 import org.wordpress.android.fluxc.network.rest.wpapi.WPAPINetworkError
-import org.wordpress.android.fluxc.network.rest.wpapi.WPAPIResponse
+import org.wordpress.android.fluxc.network.rest.wpapi.WPAPIResponse.Error
+import org.wordpress.android.fluxc.network.rest.wpapi.WPAPIResponse.Success
```
np, just a personal taste, I prefer the verbose import for its clarity of the type of the nested class.

hichamboushaba reviewed (approved):
Nice job @hafizrahman.
I'm pre-approving, I left some comments, please check them before merging.

hafizrahman committed:
@Singleton annotation not needed anymore.

See discussion  https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2608#discussion_r1057501676

hafizrahman committed:
Rename various "params" variable to "body" for better clarity.

hafizrahman committed:
Use toWooPayload() whenever possible.

hafizrahman committed:
Make Success/Error type more explicit.

hafizrahman commented:
Thanks for the review @hichamboushaba !

I've addressed all points. I could use another quick look with the `toWooPayload()` mapper usages just for some confidence check. Thanks!

hichamboushaba reviewed (approved):
Thanks for making the last changes @hafizrahman, nice work 🥇.
